lib/pbench/test/unit/server/test_user_auth.py

```diff
@@ -112,6 +114,44 @@ def test_user_login(client, server_config):
             assert response.content_type == "application/json"
             assert response.status_code == HTTPStatus.OK
 
+    @staticmethod
+    @pytest.mark.parametrize(
+        "expected_success, token_expiry",
+        [
+            (True, {"minutes": 1}),
+            (True, {"minutes": 10, "days": 1}),
+            (True, {"seconds": 1, "minutes": 1, "hours": 1, "days": 1, "weeks": 1}),
+            (True, {}),
+            (False, {"bad_key": 1}),
+            (False, {"bad_key": 1, "minutes": 1}),
+            (False, "bad_format"),
+        ],
+    )
+    def test_custom_token_expiry(client, server_config, token_expiry, expected_success):
+        """Test for login of registered-user login"""
+        with client:
+            # user registration
+            resp_register = register_user(
+                client,
+                server_config,
+                username="user",
+                firstname="firstname",
+                lastname="lastName",
+                email="user@domain.com",
+                password="12345",
+            )
+            assert resp_register.status_code == HTTPStatus.CREATED
+            # registered user login
+            response = login_user(client, server_config, "user", "12345", token_expiry)
+            data = response.json
+            if expected_success:
+                assert data["auth_token"]
```

Comment:
It is possible but when decoding the tokens I would have to hard-code the `secret_key` and `algorithm` that we used for encoding the token. After that we have to convert the time like this ` {"minutes": 10, "days": 1}` into a future epoch format or convert the decoded `exp` epoch time into a `DateTime` and then diff it with the current `DateTime`. I will do some digging into tomorrow.

Reply:
I dont think it is feasible unless I mock the `encode_auth_token` method from the `auth.py` file. I have to add the following code to make the assertion but the assertion is not guaranteed to succeed because of the involvement of `datetime.datetime.utcnow()`. We get the current UTC time when encoding the auth token so in order for the `exp` assertion to succeed we have to match that time.
```
import jwt
import os
import datetime
payload = jwt.decode(
    data["auth_token"],
    os.getenv("SECRET_KEY", "my_precious"),
    algorithms="HS256",
)
exp = datetime.datetime.utcnow() + datetime.timedelta(**token_expiry)
assert payload.get("exp") == int(exp.timestamp())
```

Reply:
If the value of the current time is the big issue, I think that that is easily dealt with:  elsewhere in the unit tests, we use a "freeze gun" to control the current time and set it to a known value (and, I believe, to prevent it from changing during the test).  So, under that control, an approach like the above should work fine, I think.

Failing that, we just do the typical assertions,  checking that the expiration lands within certain bounds (i.e., not too far in the past (preferably somewhat in the future) and not too far in the future).  But, I think `freezegun.freeze_time()` will cover it.

Beyond that, is there a problem with moking `encode_auth_token()`?